godel_process_planning/src/common_utils.cpp

```diff
@@ -0,0 +1,239 @@
+#include "common_utils.h"
+#include <eigen_conversions/eigen_msg.h>
+#include <descartes_planner/dense_planner.h>
+#include <descartes_planner/sparse_planner.h>
+#include <descartes_trajectory/axial_symmetric_pt.h>
+
+#include <moveit/kinematic_constraints/utils.h>
+#include <moveit_msgs/GetMotionPlan.h>
+
+#include <ros/topic.h>
+#include "trajectory_utils.h"
+
+Eigen::Affine3d godel_process_planning::createNominalTransform(const geometry_msgs::Pose& ref_pose,
+                                                               const geometry_msgs::Point& pt)
+{
+  Eigen::Affine3d eigen_pose;
+  Eigen::Vector3d eigen_pt;
+
+  tf::poseMsgToEigen(ref_pose, eigen_pose);
+  tf::pointMsgToEigen(pt, eigen_pt);
+
+  // Translation transform
+  Eigen::Affine3d to_point;
+  to_point = Eigen::Translation3d(eigen_pt);
+
+  // Reverse the Z axis
+  Eigen::Affine3d flip_z;
+  flip_z = Eigen::AngleAxisd(M_PI, Eigen::Vector3d::UnitY());
+
+  return eigen_pose * to_point * flip_z;
+}
+
+bool godel_process_planning::descartesSolve(const godel_process_planning::DescartesTraj& in_path,
+                                            descartes_core::RobotModelConstPtr robot_model,
+                                            godel_process_planning::DescartesTraj& out_path)
+{
+  // Create planner
+  descartes_core::PathPlannerBasePtr planner (new descartes_planner::DensePlanner);
+  planner->initialize(robot_model);
+
+  // Attempt to solve the trajectory
+  if (!planner->planPath(in_path))
+  {
+    ROS_ERROR("%s: Failed to plan for given trajectory.", __FUNCTION__);
+    return false;
+  }
+
+  if (!planner->getPath(out_path))
+  {
+    ROS_ERROR("%s: Failed to retrieve path.", __FUNCTION__);
+    return false;
+  }
+
+  return true;
+}
+
+trajectory_msgs::JointTrajectory
+godel_process_planning::toROSTrajectory(const godel_process_planning::DescartesTraj& solution,
+                                        const descartes_core::RobotModel& model)
+{
+  ros::Duration from_start(0.0);
+  std::vector<double> joint_point;
+  trajectory_msgs::JointTrajectory ros_trajectory; // result
+
+  for (std::size_t i = 0; i < solution.size(); ++i)
+  {
+    solution[i]->getNominalJointPose(std::vector<double>(), model, joint_point);
```

Comment:
An empty array should be created outside the for loop and passed into the *getNominalJointPose()* method in order to avoid recreating it on each iteration.